### PR TITLE
[NFC] Refactor planStandardCompile

### PR DESCRIPTION
**Pull request by davidungar**
Prior to work on incremental compilation, refactor `planStandardCompile` in order to clarify causality chains.
Does change the order of planned jobs in `jobs`, but I believe that isn't the scheduled order, anyway.

Passing some mutable state around via "add" functions to clarify that the callee is only adding to the collections in question.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
@swift-ci please test

File: Sources/SwiftDriver/Jobs/Planning.swift [hunk 1]
```diff
@@ -27,235 +27,349 @@ public enum PlanningError: Error, DiagnosticData {
...
 
-      partitions = nil
-
-    case .standardCompile:
-      partitions = nil
+  private mutating func addJobsForPrimaryInputs(
+    addJob: (Job) -> Void,
+    addModuleInput: (TypedVirtualPath) -> Void,
+    addLinkerInput: (TypedVirtualPath) -> Void,
+    addJobOutputs: ([TypedVirtualPath]) -> Void)
```
Comment (DougGregor) [inline Sources/SwiftDriver/Jobs/Planning.swift:194]: It feels odd that we have a mix of singular and plural `add*` functions, including having both `addLinkerInput` and `addLinkerInputs`. Can we standardize on the singular, perhaps?

File: Sources/SwiftDriver/Jobs/Planning.swift [hunk 2]
```diff
@@ -27,235 +27,349 @@ public enum PlanningError: Error, DiagnosticData {
...
 
+    var jobs = [Job]()
     // Keep track of the various outputs we care about from the jobs we build.
+
+    // Passing in these functions instead of inout parameters clarifies that
+    // the code is only adding to the collections in question.
+    func addJob(_ j: Job) {
+      jobs.append(j)
+    }
+    func addJobs(_ js: [Job]) {
```
Comment (DougGregor) [inline Sources/SwiftDriver/Jobs/Planning.swift:46]: I note this elsewhere, but I think it's awkward to have both the singular and plural forms of these. Let's standardize on one.

**Review by DougGregor — COMMENTED** [review]
Some small requests.

**Review by DougGregor — APPROVED** [review]

File: Sources/SwiftDriver/Jobs/Planning.swift [hunk 3]
```diff
@@ -27,235 +27,349 @@ public enum PlanningError: Error, DiagnosticData {
...
   /// primary files.
   private mutating func planStandardCompile() throws -> [Job] {
-    var jobs = [Job]()
+    precondition(compilerMode.isStandardCompilationForPlanning,
+                 "compiler mode \(compilerMode) is handled elsewhere")
 
+    var jobs = [Job]()
     // Keep track of the various outputs we care about from the jobs we build.
+
+    // Passing in these functions instead of inout parameters clarifies that
```
[the remark this answers is not in the document]
Comment (davidungar) [reply Sources/SwiftDriver/Jobs/Planning.swift:41]: Thanks, done! (as soon as I push)

File: Sources/SwiftDriver/Jobs/Planning.swift (hunk 1 above)
Comment (davidungar) [reply Sources/SwiftDriver/Jobs/Planning.swift:186]: Yes, good idea. Done!

File: Sources/SwiftDriver/Jobs/Planning.swift (hunk 2 above)
Comment (davidungar) [reply Sources/SwiftDriver/Jobs/Planning.swift]: Thanks done!

**Comment by davidungar** [comment]
Thanks for your review. Great suggestions, and I've taken them.

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
@swift-ci smoke test

**Comment by davidungar** [comment]
@swift-ci please test

**Pull request merged**